File: frontend/lib/features/quests/data/player_quest.dart
import 'package:equatable/equatable.dart';

/// Display states the backend computes on each /quests/me sync. The
/// persisted Domain state is only Active|Claimed, but the read pass
/// promotes Active rows to ReadyToClaim when progress >= threshold.
enum QuestState {
  active,
  readyToClaim,
  claimed,
  unknown;

  static QuestState fromString(String value) {
    switch (value) {
      case 'Active':
        return QuestState.active;
      case 'ReadyToClaim':
        return QuestState.readyToClaim;
      case 'Claimed':
        return QuestState.claimed;
      default:
        return QuestState.unknown;
    }
  }
}

class PlayerQuest extends Equatable {
  const PlayerQuest({
    required this.id,
    required this.playerId,
    required this.questDefinitionId,
    required this.name,
    required this.description,
    required this.trigger,
    required this.state,
    required this.progress,
    required this.threshold,
    required this.energyReward,
    required this.hintReward,
    required this.undoReward,
    required this.resetReward,
    required this.issuedAt,
    required this.claimedAt,
    required this.expiresAt,
  });

  factory PlayerQuest.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final playerId = json['playerId'];
    final questDefinitionId = json['questDefinitionId'];
    final name = json['name'];
    final description = json['description'];
    final trigger = json['trigger'];
    final displayState = json['displayState'];
    final progress = json['progress'];
    final threshold = json['threshold'];
    final energyReward = json['energyReward'];
    final hintReward = json['hintReward'];
    final undoReward = json['undoReward'];
    final resetReward = json['resetReward'];
    final issuedAt = json['issuedAt'];
    final claimedAt = json['claimedAt'];
    final expiresAt = json['expiresAt'];

    if (id is! String ||
        id.isEmpty ||
        playerId is! String ||
        playerId.isEmpty ||
        questDefinitionId is! String ||
        questDefinitionId.isEmpty ||
        name is! String ||
        name.isEmpty ||
        trigger is! String ||
        trigger.isEmpty ||
        displayState is! String ||
        displayState.isEmpty ||
        progress is! int ||
        threshold is! int ||
        energyReward is! int ||
        hintReward is! int ||
        undoReward is! int ||
        resetReward is! int ||
        issuedAt is! String ||
        (description != null && description is! String) ||
        (claimedAt != null && claimedAt is! String) ||
        (expiresAt != null && expiresAt is! String)) {
      throw StateError('Player quest response is missing required fields.');
    }

    return PlayerQuest(
      id: id,
      playerId: playerId,
      questDefinitionId: questDefinitionId,
      name: name,
      description: (description as String?) ?? '',
      trigger: trigger,
      state: QuestState.fromString(displayState),
      progress: progress,
      threshold: threshold,
      energyReward: energyReward,
      hintReward: hintReward,
      undoReward: undoReward,
      resetReward: resetReward,
      issuedAt: DateTime.parse(issuedAt),
      claimedAt: claimedAt == null
          ? null
          : DateTime.parse(claimedAt as String),
      expiresAt: expiresAt == null
          ? null
          : DateTime.parse(expiresAt as String),
    );
  }

  final String id;
  final String playerId;
  final String questDefinitionId;
  final String name;
  final String description;
  final String trigger;
  final QuestState state;
  final int progress;
  final int threshold;
  final int energyReward;
  final int hintReward;
  final int undoReward;
  final int resetReward;
  final DateTime issuedAt;
  final DateTime? claimedAt;
  final DateTime? expiresAt;

  bool get isReadyToClaim => state == QuestState.readyToClaim;
  bool get isClosed => state == QuestState.claimed;

  @override
  List<Object?> get props => [
    id,
    playerId,
    questDefinitionId,
    name,
    description,
    trigger,
    state,
    progress,
    threshold,
    energyReward,
    hintReward,
    undoReward,
    resetReward,
    issuedAt,
    claimedAt,
    expiresAt,
  ];
}

File: src/Modules/Quests/Domain/PlayerQuests/PlayerQuest.cs
using LexiLink.Common.Domain;
using LexiLink.Modules.Quests.Domain.PlayerQuests.Events;
using LexiLink.Modules.Quests.Domain.PlayerQuests.Rules;

namespace LexiLink.Modules.Quests.Domain.PlayerQuests;

/// <summary>
/// A player's instance of a <see cref="QuestDefinition"/>. Post Sprint Q1
/// the row stores only what cannot be derived from Stats: the snapshot
/// of the player's counter at issuance, the claim transition timestamp,
/// and (for daily quests) the expiry deadline. Progress and
/// ReadyToClaim are recomputed at every read from
/// <c>counter - ProgressBaselineSnapshot</c>.
/// </summary>
public class PlayerQuest : Entity, IAggregateRoot
{
    public PlayerQuestId Id { get; private set; }

    private Guid _playerId;
    private QuestDefinitionId _questDefinitionId = null!;
    private int _progressBaselineSnapshot;
    private QuestState _state;
    private DateTime _issuedAt;
    private DateTime? _claimedAt;
    private DateTime? _expiresAt;

    public Guid PlayerId => _playerId;
    public QuestDefinitionId QuestDefinitionId => _questDefinitionId;
    public int ProgressBaselineSnapshot => _progressBaselineSnapshot;
    public QuestState State => _state;
    public DateTime IssuedAt => _issuedAt;
    public DateTime? ClaimedAt => _claimedAt;
    public DateTime? ExpiresAt => _expiresAt;

    private PlayerQuest()
    {
        Id = null!;
    }

    private PlayerQuest(
        PlayerQuestId id,
        Guid playerId,
        QuestDefinitionId questDefinitionId,
        int progressBaselineSnapshot,
        DateTime issuedAt,
        DateTime? expiresAt)
    {
        Id = id;
        _playerId = playerId;
        _questDefinitionId = questDefinitionId;
        _progressBaselineSnapshot = Math.Max(0, progressBaselineSnapshot);
        _state = QuestState.Active;
        _issuedAt = issuedAt;
        _expiresAt = expiresAt;

        AddDomainEvent(new PlayerQuestIssuedDomainEvent(Id, _playerId, _questDefinitionId));
    }

    internal static PlayerQuest IssueFor(
        Guid playerId,
        QuestDefinitionId questDefinitionId,
        int progressBaselineSnapshot,
        DateTime issuedAt,
        DateTime? expiresAt = null)
    {
        return new PlayerQuest(
            new PlayerQuestId(Guid.NewGuid()),
            playerId,
            questDefinitionId,
            progressBaselineSnapshot,
            issuedAt,
            expiresAt);
    }

    /// <summary>
    /// Transition Active → Claimed. The caller computes
    /// <paramref name="isReadyToClaim"/> from the current Stats counter
    /// (<c>counter - <see cref="ProgressBaselineSnapshot"/> &gt;=
    /// QuestDefinition.Threshold</c>) and the caller has also verified
    /// the row is not already past <see cref="ExpiresAt"/>. Reward
    /// values live on <see cref="QuestDefinition"/> and are carried
    /// into the claimed domain event by the handler so Energy, Hint,
    /// Undo, and Reset can each grant their portion event-driven.
    /// </summary>
    internal void Claim(
        DateTime now,
        bool isReadyToClaim,
        int energyReward,
        int hintReward,
        int undoReward,
        int resetReward)
    {
        CheckRule(new QuestMustBeReadyToBeClaimedRule(_state, isReadyToClaim));

        _state = QuestState.Claimed;
        _claimedAt = now;

        AddDomainEvent(new PlayerQuestClaimedDomainEvent(
            Id,
            _playerId,
            _questDefinitionId,
            energyReward,
            hintReward,
            undoReward,
            resetReward));
    }
}
